apps/api/src/app/workflows-v2/usecases/get-step-schema/get-step-metadata-use-case.service.ts

```diff
@@ -0,0 +1,150 @@
+import { BadRequestException, Injectable } from '@nestjs/common';
+import { ControlValuesLevelEnum, StepMetadataDto, StepType, StepTypeEnum } from '@novu/shared';
+import { JSONSchema } from 'json-schema-to-ts';
+import { ControlValuesRepository, NotificationStepEntity, NotificationTemplateRepository } from '@novu/dal';
+import { GetStepSchemaCommand } from './get-step-schema.command';
+import { mapStepTypeToResult } from '../../shared';
+
+@Injectable()
+export class GetStepMetadataUseCase {
+  constructor(
+    private notificationTemplateRepository: NotificationTemplateRepository,
+    private controlValuesRepository: ControlValuesRepository
+  ) {}
+
+  async execute(command: GetStepSchemaCommand): Promise<StepMetadataDto> {
+    const { currentStep, previousSteps } = await this.findSteps(command);
+    this.assertStepValidity(currentStep);
+
+    return {
+      controls: {
+        dataSchema: currentStep.controls?.schema,
+        uiSchema: currentStep.controls?.uiSchema,
+        values: await this.getValues(command, currentStep),
+      },
+      variables: buildVariablesSchema(previousSteps),
+    };
+  }
+
+  private async getValues(command: GetStepSchemaCommand, currentStep: NotificationStepEntity) {
+    const controlValuesEntity = await this.controlValuesRepository.findFirst({
+      _environmentId: command.environmentId,
+      _organizationId: command.organizationId,
+      _workflowId: command.workflowId,
+      _stepId: currentStep._templateId,
+      level: ControlValuesLevelEnum.STEP_CONTROLS,
+    });
+
+    return controlValuesEntity?.controls || {};
+  }
+
+  private assertStepValidity(currentStep: NotificationStepEntity) {
+    if (!currentStep.template?.type) {
```

Comment:
Please move this to isValidStepType function. A step without type is invalid so one function should check that.